File: source/ts/src/jyotish/pathuPorutham.ts
import {
  RASHI_LORD, NAISARGIKA_MAITRI,
  RASHI_VASHYA, VASHYA_SCORE, vashyaIndex,
  NAKSHATRA_YONI, YONI_SCORE, yoniIndex,
  NAKSHATRA_GANA,
} from './matchingTables';
import {
  NAKSHATRA_RAJJU, vedhaOf,
  MAHENDRA_AUSPICIOUS_DISTANCES, DINA_AUSPICIOUS_REMAINDERS,
  RASHI_DOSHIC_DISTANCES,
} from './pathuPoruthamTables';
import type { NatalMoon } from './matching';
import { assertNakshatraIndex } from '../utils/validation';

export type PoruthamName =
  | 'Dina'
  | 'Gana'
  | 'Mahendra'
  | 'SthreeDeergha'
  | 'Yoni'
  | 'Rashi'
  | 'Rashyathipathi'
  | 'Vasya'
  | 'Rajju'
  | 'Vedha';

export interface PoruthamScore {
  name: PoruthamName;
  passes: boolean;
  description: string;
  /** Present only when this koot is a strong veto (Yoni / Rajju / Vedha) and failed. */
  veto?: boolean;
}

export interface PathuPoruthamResult {
  totalPasses: number;
  /** True iff no veto-level koot failed and `totalPasses >= 5`. */
  recommended: boolean;
  poruthams: PoruthamScore[];
}

/**
 * Pathu Porutham, the Tamil 10-fold pass/fail marriage compatibility test; Dina,
 * Mahendra and Sthree Deergha count nakshatra distance girl to boy, per Tamil
 * convention.
 */
export function computePathuPorutham(boy: NatalMoon, girl: NatalMoon): PathuPoruthamResult {
  validateNatalMoon(boy, 'boy');
  validateNatalMoon(girl, 'girl');

  const poruthams: PoruthamScore[] = [
    scoreDina(boy, girl),
    scoreGana(boy, girl),
    scoreMahendra(boy, girl),
    scoreSthreeDeergha(boy, girl),
    scoreYoni(boy, girl),
    scoreRashi(boy, girl),
    scoreRashyathipathi(boy, girl),
    scoreVasya(boy, girl),
    scoreRajju(boy, girl),
    scoreVedha(boy, girl),
  ];

  const totalPasses = poruthams.reduce((c, k) => c + (k.passes ? 1 : 0), 0);
  const vetoFailed = poruthams.some((k) => k.veto === true);
  const recommended = !vetoFailed && totalPasses >= 5;

  return { totalPasses, recommended, poruthams };
}

function validateNatalMoon(m: NatalMoon, label: string): void {
  if (!Number.isInteger(m.rashi) || m.rashi < 0 || m.rashi >= 12) {
    throw new RangeError(`${label}.rashi must be integer in [0, 11], got ${m.rashi}`);
  }
  assertNakshatraIndex(m.nakshatra, `${label}.nakshatra`);
  if (m.lagnaRashi !== undefined &&
      (!Number.isInteger(m.lagnaRashi) || m.lagnaRashi < 0 || m.lagnaRashi >= 12)) {
    throw new RangeError(`${label}.lagnaRashi must be integer in [0, 11], got ${m.lagnaRashi}`);
  }
  if (m.navamsaRashi !== undefined &&
      (!Number.isInteger(m.navamsaRashi) || m.navamsaRashi < 0 || m.navamsaRashi >= 12)) {
    throw new RangeError(`${label}.navamsaRashi must be integer in [0, 11], got ${m.navamsaRashi}`);
  }
  if (m.nakshatraPada !== undefined &&
      (!Number.isInteger(m.nakshatraPada) || m.nakshatraPada < 1 || m.nakshatraPada > 4)) {
    throw new RangeError(`${label}.nakshatraPada must be integer in [1, 4], got ${m.nakshatraPada}`);
  }
}

function scoreDina(boy: NatalMoon, girl: NatalMoon): PoruthamScore {
  // Remainder 1 (Janma) is Mixed in the 9-Tara scheme but fails in this binary one.
  const distance = ((boy.nakshatra - girl.nakshatra + 27) % 27) + 1;
  const remainder = distance % 9;
  const passes = DINA_AUSPICIOUS_REMAINDERS.includes(remainder);
  return {
    name: 'Dina',
    passes,
    description: `Girl→Boy nakshatra distance ${distance} (mod 9 = ${remainder}), ${passes ? 'auspicious' : 'inauspicious'}`,
  };
}

function scoreGana(boy: NatalMoon, girl: NatalMoon): PoruthamScore {
  const boyGana = NAKSHATRA_GANA[boy.nakshatra]!;
  const girlGana = NAKSHATRA_GANA[girl.nakshatra]!;
  // Only Manushya-Rakshasa fails; Deva-Rakshasa counts as a pass per AstroVed.
  const fails =
    (boyGana === 'manushya' && girlGana === 'rakshasa') ||
    (boyGana === 'rakshasa' && girlGana === 'manushya');
  return {
    name: 'Gana',
    passes: !fails,
    description: `Boy ${boyGana} ↔ Girl ${girlGana}${fails ? ', clash' : ''}`,
  };
}

function scoreMahendra(boy: NatalMoon, girl: NatalMoon): PoruthamScore {
  // The Mahendra set is closed under N → 29-N, so the verdict is direction-independent.
  const distance = ((boy.nakshatra - girl.nakshatra + 27) % 27) + 1;
  const passes = MAHENDRA_AUSPICIOUS_DISTANCES.includes(distance);
  return {
    name: 'Mahendra',
    passes,
    description: `Girl→Boy nakshatra distance ${distance}, ${passes ? 'in Mahendra set {4,7,10,13,16,19,22,25}' : 'not in Mahendra set'}`,
  };
}

function scoreSthreeDeergha(boy: NatalMoon, girl: NatalMoon): PoruthamScore {
  // Threshold locked at > 13 ("Uthamam"), the dominant Tamil consensus and the
  // one the reference almanac follows.
  const distance = ((boy.nakshatra - girl.nakshatra + 27) % 27) + 1;
  const passes = distance > 13;
  return {
    name: 'SthreeDeergha',
    passes,
    description: `Girl→Boy nakshatra distance ${distance}, ${passes ? 'longevity favorable (> 13)' : 'longevity weak (≤ 13)'}`,
  };
}

function scoreYoni(boy: NatalMoon, girl: NatalMoon): PoruthamScore {
  const boyYoni = NAKSHATRA_YONI[boy.nakshatra]!;
  const girlYoni = NAKSHATRA_YONI[girl.nakshatra]!;
  const score = YONI_SCORE[yoniIndex(boyYoni)]![yoniIndex(girlYoni)]!;
  // 0 = enemy yoni → veto; 1 (unfriendly) fails but does not veto.
  const passes = score >= 2;
  const veto = score === 0;
  const out: PoruthamScore = {
    name: 'Yoni',
    passes,
    description: `${boyYoni} ↔ ${girlYoni} (Ashtakoot Yoni score ${score}/4)`,
  };
  if (veto) out.veto = true;
  return out;
}

function scoreRashi(boy: NatalMoon, girl: NatalMoon): PoruthamScore {
  const dBoyToGirl = ((girl.rashi - boy.rashi + 12) % 12) + 1;
  const dGirlToBoy = ((boy.rashi - girl.rashi + 12) % 12) + 1;
  const isDoshic = RASHI_DOSHIC_DISTANCES.some(
    ([a, b]) => a === dBoyToGirl && b === dGirlToBoy,
  );
  return {
    name: 'Rashi',
    passes: !isDoshic,
    description: `Rashi distance (${dBoyToGirl}, ${dGirlToBoy})${isDoshic ? ', doshic' : ''}`,
  };
}

function scoreRashyathipathi(boy: NatalMoon, girl: NatalMoon): PoruthamScore {
  const boyLord = RASHI_LORD[boy.rashi]!;
  const girlLord = RASHI_LORD[girl.rashi]!;
  if (boyLord === girlLord) {
    return {
      name: 'Rashyathipathi',
      passes: true,
      description: `Same rashi-lord (graha index ${boyLord}), full compatibility`,
    };
  }
  const boyView = NAISARGIKA_MAITRI[boyLord]![girlLord]!;
  const girlView = NAISARGIKA_MAITRI[girlLord]![boyLord]!;
  const passes = boyView !== -1 && girlView !== -1;
  return {
    name: 'Rashyathipathi',
    passes,
    description: `Boy lord views girl lord as ${maitriLabel(boyView)}; girl lord views boy lord as ${maitriLabel(girlView)}`,
  };
}

function maitriLabel(v: number): string {
  return v === 1 ? 'friend' : v === -1 ? 'enemy' : 'neutral';
}

function scoreVasya(boy: NatalMoon, girl: NatalMoon): PoruthamScore {
  const boyVashya = RASHI_VASHYA[boy.rashi]!;
  const girlVashya = RASHI_VASHYA[girl.rashi]!;
  const score = VASHYA_SCORE[vashyaIndex(boyVashya)]![vashyaIndex(girlVashya)]!;
  const passes = score > 0;
  return {
    name: 'Vasya',
    passes,
    description: `${boyVashya} ↔ ${girlVashya} (Ashtakoot Vashya score ${score}/2)`,
  };
}

function scoreRajju(boy: NatalMoon, girl: NatalMoon): PoruthamScore {
  const boyRajju = NAKSHATRA_RAJJU[boy.nakshatra]!;
  const girlRajju = NAKSHATRA_RAJJU[girl.nakshatra]!;
  const sameRajju = boyRajju === girlRajju;
  const out: PoruthamScore = {
    name: 'Rajju',
    passes: !sameRajju,
    description: sameRajju
      ? `Both in ${boyRajju} Rajju, strong veto (longevity threat)`
      : `Boy ${boyRajju} ↔ Girl ${girlRajju}, different rajju`,
  };
  if (sameRajju) out.veto = true;
  return out;
}

function scoreVedha(boy: NatalMoon, girl: NatalMoon): PoruthamScore {
  const partner = vedhaOf(boy.nakshatra);
  const isVedha = partner !== null && partner === girl.nakshatra;
  const out: PoruthamScore = {
    name: 'Vedha',
    passes: !isVedha,
    description: isVedha
      ? `Boy nakshatra ${boy.nakshatra} ↔ Girl nakshatra ${girl.nakshatra} are vedha partners, strong veto`
      : 'No Vedha obstruction',
  };
  if (isVedha) out.veto = true;
  return out;
}

File: source/go/internal/jyotish/pathuporutham.go
package jyotish

import (
	"github.com/ishankgupta95/panchang-ts/source/go/v5/internal/jsnum"
)

type PoruthamName string

const (
	PoruthamDina           PoruthamName = "Dina"
	PoruthamGana           PoruthamName = "Gana"
	PoruthamMahendra       PoruthamName = "Mahendra"
	PoruthamSthreeDeergha  PoruthamName = "SthreeDeergha"
	PoruthamYoni           PoruthamName = "Yoni"
	PoruthamRashi          PoruthamName = "Rashi"
	PoruthamRashyathipathi PoruthamName = "Rashyathipathi"
	PoruthamVasya          PoruthamName = "Vasya"
	PoruthamRajju          PoruthamName = "Rajju"
	PoruthamVedha          PoruthamName = "Vedha"
)

var AllPoruthamNames = [10]PoruthamName{
	PoruthamDina, PoruthamGana, PoruthamMahendra, PoruthamSthreeDeergha,
	PoruthamYoni, PoruthamRashi, PoruthamRashyathipathi, PoruthamVasya,
	PoruthamRajju, PoruthamVedha,
}

var VetoPoruthams = [3]PoruthamName{PoruthamYoni, PoruthamRajju, PoruthamVedha}

type PoruthamScore struct {
	Name        PoruthamName `json:"name"`
	Passes      bool         `json:"passes"`
	Description string       `json:"description"`
	Veto        *bool        `json:"veto,omitempty"`
}

type PathuPoruthamResult struct {
	TotalPasses int             `json:"totalPasses"`
	Recommended bool            `json:"recommended"`
	Poruthams   []PoruthamScore `json:"poruthams"`
}

func ComputePathuPorutham(boy, girl NatalMoon) (PathuPoruthamResult, error) {
	if err := validateNatalMoon(boy, "boy"); err != nil {
		return PathuPoruthamResult{}, err
	}
	if err := validateNatalMoon(girl, "girl"); err != nil {
		return PathuPoruthamResult{}, err
	}

	poruthams := []PoruthamScore{
		scoreDina(boy, girl),
		scorePoruthamGana(boy, girl),
		scoreMahendra(boy, girl),
		scoreSthreeDeergha(boy, girl),
		scorePoruthamYoni(boy, girl),
		scoreRashi(boy, girl),
		scoreRashyathipathi(boy, girl),
		scoreVasya(boy, girl),
		scoreRajju(boy, girl),
		scoreVedha(boy, girl),
	}

	totalPasses := 0
	vetoFailed := false
	for _, k := range poruthams {
		if k.Passes {
			totalPasses++
		}
		if k.Veto != nil && *k.Veto {
			vetoFailed = true
		}
	}

	return PathuPoruthamResult{
		TotalPasses: totalPasses,
		Recommended: !vetoFailed && totalPasses >= 5, // AstroVed / reference-almanac threshold
		Poruthams:   poruthams,
	}, nil
}

func vetoTrue() *bool { t := true; return &t }

func scoreDina(boy, girl NatalMoon) PoruthamScore {
	distance := ((boy.Nakshatra-girl.Nakshatra+27)%27 + 1)
	remainder := distance % 9
	passes := false
	for _, r := range DinaAuspiciousRemainders {
		if r == remainder {
			passes = true
			break
		}
	}
	return PoruthamScore{
		Name: PoruthamDina, Passes: passes,
		Description: "Girl→Boy nakshatra distance " + jsnum.FormatInt(int64(distance)) +
			" (mod 9 = " + jsnum.FormatInt(int64(remainder)) + "), " + auspiciousLabel(passes),
	}
}

func scorePoruthamGana(boy, girl NatalMoon) PoruthamScore {
	boyGana := NakshatraGana[boy.Nakshatra]
	girlGana := NakshatraGana[girl.Nakshatra]
	fails := (boyGana == GanaManushya && girlGana == GanaRakshasa) ||
		(boyGana == GanaRakshasa && girlGana == GanaManushya)
	description := "Boy " + string(boyGana) + " ↔ Girl " + string(girlGana)
	if fails {
		description += ", clash"
	}
	return PoruthamScore{Name: PoruthamGana, Passes: !fails, Description: description}
}

func scoreMahendra(boy, girl NatalMoon) PoruthamScore {
	distance := ((boy.Nakshatra-girl.Nakshatra+27)%27 + 1)
	passes := false
	for _, d := range MahendraAuspiciousDistances {
		if d == distance {
			passes = true
			break
		}
	}
	tail := "not in Mahendra set"
	if passes {
		tail = "in Mahendra set {4,7,10,13,16,19,22,25}"
	}
	return PoruthamScore{
		Name: PoruthamMahendra, Passes: passes,
		Description: "Girl→Boy nakshatra distance " + jsnum.FormatInt(int64(distance)) + ", " + tail,
	}
}

func scoreSthreeDeergha(boy, girl NatalMoon) PoruthamScore {
	distance := ((boy.Nakshatra-girl.Nakshatra+27)%27 + 1)
	passes := distance > 13
	tail := "longevity weak (≤ 13)"
	if passes {
		tail = "longevity favorable (> 13)"
	}
	return PoruthamScore{
		Name: PoruthamSthreeDeergha, Passes: passes,
		Description: "Girl→Boy nakshatra distance " + jsnum.FormatInt(int64(distance)) + ", " + tail,
	}
}

func scorePoruthamYoni(boy, girl NatalMoon) PoruthamScore {
	boyYoni := NakshatraYoni[boy.Nakshatra]
	girlYoni := NakshatraYoni[girl.Nakshatra]
	score := YoniScore[YoniIndex(boyYoni)][YoniIndex(girlYoni)]
	out := PoruthamScore{
		Name: PoruthamYoni, Passes: score >= 2,
		Description: string(boyYoni) + " ↔ " + string(girlYoni) +
			" (Ashtakoot Yoni score " + jsnum.FormatInt(int64(score)) + "/4)",
	}
	if score == 0 {
		out.Veto = vetoTrue()
	}
	return out
}

func scoreRashi(boy, girl NatalMoon) PoruthamScore {
	dBoyToGirl := ((girl.Rashi-boy.Rashi+12)%12 + 1)
	dGirlToBoy := ((boy.Rashi-girl.Rashi+12)%12 + 1)
	isDoshic := false
	for _, p := range RashiDoshicDistances {
		if p[0] == dBoyToGirl && p[1] == dGirlToBoy {
			isDoshic = true
			break
		}
	}
	description := "Rashi distance (" + jsnum.FormatInt(int64(dBoyToGirl)) + ", " +
		jsnum.FormatInt(int64(dGirlToBoy)) + ")"
	if isDoshic {
		description += ", doshic"
	}
	return PoruthamScore{Name: PoruthamRashi, Passes: !isDoshic, Description: description}
}

func scoreRashyathipathi(boy, girl NatalMoon) PoruthamScore {
	boyLord := RashiLord[boy.Rashi]
	girlLord := RashiLord[girl.Rashi]
	if boyLord == girlLord {
		return PoruthamScore{
			Name: PoruthamRashyathipathi, Passes: true,
			Description: "Same rashi-lord (graha index " +
				jsnum.FormatInt(int64(boyLord)) + "), full compatibility",
		}
	}
	boyView := NaisargikaMaitri[boyLord][girlLord]
	girlView := NaisargikaMaitri[girlLord][boyLord]
	return PoruthamScore{
		Name: PoruthamRashyathipathi, Passes: boyView != -1 && girlView != -1,
		Description: "Boy lord views girl lord as " + maitriLabel(boyView) +
			"; girl lord views boy lord as " + maitriLabel(girlView),
	}
}

func scoreVasya(boy, girl NatalMoon) PoruthamScore {
	boyVashya := RashiVashya[boy.Rashi]
	girlVashya := RashiVashya[girl.Rashi]
	score := VashyaScore[VashyaIndex(boyVashya)][VashyaIndex(girlVashya)]
	return PoruthamScore{
		Name: PoruthamVasya, Passes: score > 0,
		Description: string(boyVashya) + " ↔ " + string(girlVashya) +
			" (Ashtakoot Vashya score " + jsnum.FormatFloat(score) + "/2)",
	}
}

func scoreRajju(boy, girl NatalMoon) PoruthamScore {
	boyRajju := NakshatraRajju[boy.Nakshatra]
	girlRajju := NakshatraRajju[girl.Nakshatra]
	sameRajju := boyRajju == girlRajju
	description := "Boy " + string(boyRajju) + " ↔ Girl " + string(girlRajju) + ", different rajju"
	if sameRajju {
		description = "Both in " + string(boyRajju) + " Rajju, strong veto (longevity threat)"
	}
	out := PoruthamScore{Name: PoruthamRajju, Passes: !sameRajju, Description: description}
	if sameRajju {
		out.Veto = vetoTrue()
	}
	return out
}

func scoreVedha(boy, girl NatalMoon) PoruthamScore {
	partner, hasPartner := VedhaOf(boy.Nakshatra)
	isVedha := hasPartner && partner == girl.Nakshatra
	description := "No Vedha obstruction"
	if isVedha {
		description = "Boy nakshatra " + jsnum.FormatInt(int64(boy.Nakshatra)) +
			" ↔ Girl nakshatra " + jsnum.FormatInt(int64(girl.Nakshatra)) +
			" are vedha partners, strong veto"
	}
	out := PoruthamScore{Name: PoruthamVedha, Passes: !isVedha, Description: description}
	if isVedha {
		out.Veto = vetoTrue()
	}
	return out
}
